src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs

```diff
@@ -0,0 +1,416 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.BraceCompletion;
+using Microsoft.CodeAnalysis.CSharp.Formatting;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Formatting.Rules;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Indentation;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.BraceCompletion
+{
+    [Export(LanguageNames.CSharp, typeof(IBraceCompletionService)), Shared]
+    internal class CurlyBraceCompletionService : AbstractBraceCompletionService
+    {
+        private static readonly SyntaxAnnotation s_closingBraceSyntaxAnnotation = new("original closing brace");
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CurlyBraceCompletionService()
+        {
+        }
+
+        protected override char OpeningBrace => CurlyBrace.OpenCharacter;
+
+        protected override char ClosingBrace => CurlyBrace.CloseCharacter;
+
+        public override async Task<bool> AllowOverTypeAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+        {
+            return await CheckCurrentPositionAsync(context.Document, context.CaretLocation, cancellationToken).ConfigureAwait(false)
+                && await CheckClosingTokenKindAsync(context.Document, context.ClosingPoint, cancellationToken).ConfigureAwait(false);
```

Comment:
shared logic with another service.  should tehre be a well-named helper that expresses this idea that we can doc and which can be used from all the similar services?